### ISSUE TITLE
Change table headings in mxCompare output to support WLS

### ISSUE THREAD

**Issue by tbates** [OpenMx/OpenMx#297]
For WLS `mxCompare`  should show `model@output$fit`  not the contents of `$Minus2LogLikelihood`

The relevant columns in mxCompare output are labels "LL" and "diffLL"

|base|comparison|ep|minus2LL|df|AIC|diffLL|diffdf|p|
|--|--|--|--|--|--|--|--|--|--|
| - | - | - | - | - | - | - | - | - | - |

One option for display would be to relabel the table, and maybe add a fitUnits column.

|base|comparison|ep|fit|df|AIC|diffFit|diffdf|p|fitUnits
|--|--|--|--|--|--|--|--|--|--|
| - | - | - | - | - | - | - | - | - | - |

This would break code consuming `mxCompare` output, but guessing that's only `umx`, and I can quickly fix.

An alternative would be to leave the columns as-is and print a note after the table when WLS models are compared, politely noting that the LL columns contain  `r'Wr` units and define these (not sure that's defined anywhere?)

based on 
https://github.com/OpenMx/OpenMx/issues/289#issuecomment-682214735_

**Comment by mhunter1** [comment]
Making progress on this:

268e5fdd3ef3c485f47e3875eda6a248e3019c66

24035eda976bf644a0e45f660ee3a63718c9f3b6

49e62c023ce9c3baceee567b7d7efefeb241ad63

**Comment by tbates** [comment]
Great progress!

Comparing two single group models, the `SBchisq`  column is printed, but is NaN, as is the p column. regular `chisq` is also printed (is that by design and in place of -2LL?) 

```R
mxCompare(w1, w2)
  base comparison ep    chisq df      AIC SBchisq diffdf   p
1 DWLS       <NA>  6 0.000000  0 12.00000      NA     NA  NA
2 DWLS drop_l2mpg  5 4.495419  1 14.49542     NaN      1 NaN
```

Trying to compare multi-group WLS models using  `mxCompare`, results seems similar: `chisq` column filled in. But Satora-Bentler ChiSq is NaN, as is p-value.

```R
mxCompare(w1,w2)
  base comparison ep    chisq df      AIC SBchisq diffdf   p
1 DWLS       <NA>  6 0.000000  0 12.00000      NA     NA  NA
2 DWLS drop_l2mpg  5 4.495419  1 14.49542     NaN      1 NaN
```

A delta-fit column is not printed - feels like it would be informative?

In a data.frame version of the output, the diffFit column appears empty?

```R
as.data.frame(tmp)
  base comparison ep minus2LL df      AIC diffLL diffdf   p      fit fitUnits diffFit    chisq SBchisq
1 DWLS       <NA>  6       NA  0 12.00000     NA     NA  NA 0.000000     r'Wr      NA 0.000000      NA
2 DWLS drop_l2mpg  5       NA  1 14.49542     NA      1 NaN 4.495419     r'Wr     NaN 4.495419     NaN
```


model code

```R
# make a litres variable in built-in mtcars dataframe
mtcars$litres = mtcars$disp/61.02

w1 = umxRAM(data = mtcars, type="DWLS", "#DWLS 
   mpg ~ wt + litres"
)
w2 = umxModify(w1, update="litres_to_mpg", name="drop_l2mpg")
mxCompare(w1, w2)


# Build multi-group DWLS models grouping on transmission type
w1 = umxRAM("DWLS", data = mtcars, group = "am", type="DWLS",
	umxPath(c("wt", "litres"), to = "mpg"),
	umxPath("wt", with = "litres"),
	umxPath(var = c("wt", "litres", "mpg"))
)
w2 = umxModify(w1, regex=c("litres_to_mpg.*"), name="drop_l2mpg", comp=TRUE)
```

**Comment by tbates** [comment]
No p-value or fit for the above example (as of OpenMx version: 2.18.1.233 [GIT v2.18.1-233-ga849f4e3])

```R
mxCompare(w1, w2)
  base comparison ep    chisq df      AIC SBchisq diffdf   p
1 DWLS       <NA>  6 0.000000  0 12.00000      NA     NA  NA
2 DWLS drop_l2mpg  5 4.495419  1 14.49542     NaN      1 NaN
```

the ML equivalent is

```R
base comparison ep minus2LL df      AIC   diffLL diffdf          p
1   m1       <NA>  9 330.0014 87 348.0014       NA     NA         NA
2   m1 drop_l2mpg  8 333.8631 88 349.8631 3.861645      1 0.04940188
```

**Comment by mcneale** [comment]
This is at least partly deliberate, due to perceived lack of asymptotic chi-squared distribution behavior for difference in WLS fit statistics.  If you have evidence otherwise, please share!

**Comment by tbates** [comment]
Yeah, but I think the purpose of adding `SBChisq` was to enable comparison of nested models and p-values?

Satorra, A. & Bentler, P.M. (2010). Ensuring positiveness of the scaled difference chi-square test statistic. Psychometrika 75: 243. doi:10.1007/s11336-009-9135-y

**Comment by mcneale** [comment]
Correct, but if these comparisons are from the same (one hopes multivariate normal) data, there’s rather a large difference in p-values if the WLS difference is taken to be chi-squared, p~.03 vs .05.

**Comment by tbates** [comment]
So, tangential to the wider issue, there's a bug in Bug  in the C++ The standard errors and Satorra-Bentler scaling factors are calculated in ComputeStandardError::computeImpl. But if a model has an unidentified parameter or a singular information matrix the C++ code exits early. Because the scaling variables (madj, x2m, dstar, etc.) are not yet initialized and so R reads back random garbage memory values like chiMadjust = -3.8e+26 - whatever happened to be in the void :-)

Will patch

**Comment by tbates** [comment]
So eb72a604  pushes some minor changes which prevent the return of un-initialised variables. 
The rationale for these default values is to represent the identity/neutral transformations, which allows the comparison to degrade gracefully to a standard, unscaled  test when robust scaling cannot be computed.

Also pushed the parallel CI branch tbates otherwise this will get messy to work on old and new versions of code.

**Comment by tbates** [comment]
A second bug is Missing Multigroup WLS Corrections
While OpenMx computes WLS scaling factors (chiM, chiMV) for individual single-group models, it does not aggregate or calculate them for multigroup models (which use mxFitFunctionMultigroup). The container model simply receives 0 or NA for these scaling factors, making Satorra-Bentler comparisons between multigroup models impossible out-of-the-box.

**Comment by tbates** [comment]
Fixing Bug A turns out to fix Bug B.

**Comment by tbates** [comment]
Currently, running mxRefModels() on a WLS model generates the saturated and independence reference models.

But calling summary(wls_model, refModels=ref) displays NA for the model and independence fits in the Model Statistics table, and fails to compute standard fit indices (CFI, TLI, RMSEA).

This is because the summary was built to expect likelihood-based fits (-2lnL) and does not map WLS chi-square statistics to the fit fields.

Fix is to modify the summary logic so WLS chi-square statistics are correctly utilized by summary().

**Comment by tbates** [comment]
Implemented the generateReferenceModels S4 method for MxFitFunctionWLS in  R/MxFitFunctionWLS.R

Updated generateNormalReferenceModels in  R/MxSaturatedModel.R  to configure both the saturated and independence models to use the WLS fit function, adjusting mean/covariance/threshold parameters appropriately based on whether the data is continuous or ordinal/marginal

Kept standard errors and Hessian calculations enabled when building reference models for WLS, ensuring the C++ backend calculates the unscaled chi-square statistic chi.

Modified refToLikelihood and setLikelihoods in  R/MxSummary.R  to detect when a model utilizes WLS fit units (r'Wr or r'wr). When using WLS, it maps the model's unscaled chi-square (model$output$chi) to the fit fields (e.g. Minus2LogLikelihood equivalents) instead of using the likelihood values. This cleanly feeds the correct WLS chi-square statistics into the existing CFI, TLI, and RMSEA formulas without altering the mathematical equations themselves.

Running summary(wls_model, refModels=ref) now correctly computes non-NA values for CFI, TLI, and RMSEA (and confidence intervals) and prints them, and correctly prints the Model Statistics table fits without any NA values.

Added a new test file inst/models/passing/WlsRefModelsTest.R in the repository to verify the correctness of WLS reference models and fit indices.

**Comment by mhunter1** [comment]
Again, @tbates I'm not sure that any of these changes are the statistically correct approach or the computational design that the rest of the OpenMx team would agree with.
